### DISCUSSION TITLE
Microchip xc16 and section "__c30_signature"

### DISCUSSION THREAD

**Discussion by gtackett** [NationalSecurityAgency/ghidra#2646] [Q&A] [answered]
Object modules produced by the xc16 compiler, as well as those in the libraries, include a section called ```__c30_section```. (I _think_ it has something to do with debugging...)

I extracted some of the individual modules from **liblega-pic30-elf.a** and loaded a few of them individually in Ghidra. In them, I find the section ```__c30_section``` located at **ram:0062**. That's misleading, because the low addresses in **ram** actually are reserved, and **ram:0062** happens to fall within the processor's alternate interrupt vector table.

But in executable files produced by the xc16 linker, the section shows up in the address space **OTHER**.

Maybe @ghidra1 will have some ideas here?

**Comment by ghidra1** [comment]
I suggest you try setting the  base offset import option for such object modules which may not compute a good default base address.

**Reply by gtackett** [reply]
That should work for a single object module. The Microchip libraries contain thousands. I've been doing batch imports interactively, and (as probably makes sense) there's no option available there to specify a single base address for every module.

Maybe there's a headless way to import modules in bulk and specify options like base address? I'll try to figure it out myself, but a guru can probably quickly tell me how here, where it will be findable by other users.

**Comment by ghidra1** [comment, marked as answer]
You can use `-loader-imagebase` option with headless command line

**Reply by gtackett** [reply]
```-loader-dataImagebase``` is actually the one I need, since ```c30_section``` gets loaded into a memory block in```ram:```.

This does allow me to get ```__c30_section``` out of the way of the predefined registers, etc., in low memory.

Is there a way, as part of a headless import, to load a given section at a specified address and address space? 

Say, for example, I'd like ```c30_section``` to go at ```OTHER:00001000```. With a little work I could write my own script to re-create ```c30_section``` at that address, but I'd like to avoid that, if there's already an easier way.

**Reply by ghidra1** [reply]
No.  We only support a single image base for the default and data spaces.  All overlays in the OTHER space are expected to start at 0 since they do not represent a loaded section and only correspond to non-loaded file data.

**Reply by gtackett** [reply]
Sounds like time for a little scripting!
